src/nvim/popupmnu.c

```diff
@@ -537,24 +537,33 @@ void pum_redraw(void)
                 }
               }
               grid_puts_len(&pum_grid, rt, (int)STRLEN(rt), row,
-                            col - size + 1, attr);
+                            grid_col - size + 1, attr);
               xfree(rt_start);
               xfree(st);
-              col -= width;
+              grid_col -= width;
             } else {
               int size = (int)STRLEN(st);
               int cells = (int)mb_string2cells(st);
 
-              // only draw the text that fits
-              while (size > 0 && col + cells > pum_width + pum_col) {
-                size--;
-                size -= utf_head_off(st, st + size);
-                cells -= utf_ptr2cells(st + size);
+              if (size > 0 && grid_col - col_off + cells > pum_width) {
+                // only draw the text that fits
+                do {
+                  size--;
+                  size -= utf_head_off(st, st + size);
+                  cells -= utf_ptr2cells(st + size);
+                } while (size > 0 && grid_col - col_off + cells > pum_width);
+
+                if (grid_col - col_off + cells < pum_width) {
+                  // Most right character requires 2-cells but only 1 cell
+                  // is available on screen.  Put a '>' on the right of the
+                  // pum item
+                  st[size++] = '>';
```

Comment:
Tests seems be have problem with non-continuous grid, so I'm making this deviation from Vim.